### PR TITLE
Unit testing for SetNodeNameOrDie in package cmd/options

### PR DESCRIPTION
**Pull request by gkGaneshR**
1. Why is this change necessary ?
fixes: kubernetes/node-problem-detector#161

2. How does this change address the issue ?
Under package cmd/options, the testing for SetNodeNameOrDie need
to decide Nodename based on environment variable "NODE_NAME" or
hostname or hostnameoverride variable.

3. How to verify this change ?
Run "go test" with admin privilege i.e., "make test"

Signed-off-by: gkGaneshR <gkganesh126@gmail.com>

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/node-problem-detector/162)

### CODE REVIEW

**Comment by gkGaneshR** [comment]
@k8s-bot /cla

**Comment by gkGaneshR** [comment]
@k8s-ci-robot /test

**Comment by gkGaneshR** [comment]
@k8s-bot /test

File: cmd/options/options_test.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
+package options
```
Comment (andyxning) [inline cmd/options/options_test.go]: Add Copyright 2018 statement. :)

File: cmd/options/options_test.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	"testing"
+)
+
+type Options struct {
+	Nodename         string
+	Hostname         string
+	HostnameOverride string
+}
+
+//TestSetNodeNameOrDie tests for permutations of nodename, hostname and hostnameoverride
```
Comment (andyxning) [inline cmd/options/options_test.go]: Insert an blank between `//` and `TestSetNodeNameOrDie`

**Comment by gkGaneshR** [comment]
@andyxning Thanks for the review. I have made the commit making changes.

**Comment by gkGaneshR** [comment]
Can you please approve and merge the pr?

File: cmd/options/options_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+package options
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+type Options struct {
```
Comment (andyxning) [inline cmd/options/options_test.go]: IMHO, `Options` should be inner pkg level, i.e., it should be `options`. There is no need to export it.

File: cmd/options/options_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+			},
+		},
+		"Check Nodename only": {
+			Expected: Options{
+				Nodename:         "my-node-name",
+				Hostname:         "",
+				HostnameOverride: "",
+			},
+		},
+
```
Comment (andyxning) [inline cmd/options/options_test.go]: Run `gofmt -s -w cmd/options/options_test.go` to format the code.

File: cmd/options/options_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		if opt.Expected.Nodename != "" {
+			err = os.Setenv("NODE_NAME", opt.Expected.Nodename)
+			if err != nil {
+				t.Errorf("Unable to set env NODE_NAME")
+			}
+		}
+
+		if opt.Expected.Hostname != "" {
+			// Changing hostname
+			cmd := exec.Command("hostname", opt.Expected.Hostname)
```
Comment (andyxning) [inline cmd/options/options_test.go]: We should avoid setting `hostname` directly. However, setting an environment is ok to me. But you need to clear the `NODE_NAME` env for each test as the in the for loop the prior setted `NODE_NAME` environment is not cleared and will have an impact on the next test case.

Comment (andyxning) [reply cmd/options/options_test.go]: The more proper way to test hostname is we should first get the hostname and check it when we need to validate that both hostname override and NODE_NAME environment are not setted. 

The test case meta data should be 
```
map[string]struct {
		Opts         Options
                WantedNodeName string
	}
```

In each test, we should initialize `NodeProblemDetectorOptions` and base environment with `Opts` for such as setting NODE_NAME environment and hostname override. And, in the end of each test, we should check for the updated `NodeProblemDetectorOptions` instance and compare it with the desired `WantedNodeName`.

**Comment by andyxning** [comment]
@gkGaneshR Add some comments. PTAL.

**Comment by gkGaneshR** [comment]
@andyxning I have made the commit with the changes. Please take a look.

**Comment by Random-Liu** [comment]
@andyxning Thanks for reviewing! I'll merge as long as you are ok with this. :)

File: cmd/options/options_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	"os"
+	"testing"
+)
+
+type options struct {
+	Nodename         string
+	HostnameOverride string
+}
+
+// TestSetNodeNameOrDie tests for permutations of nodename, hostname and hostnameoverride
```
Comment (andyxning) [inline cmd/options/options_test.go]: Add a dot to the end of the line as per comment should be a sentence. :)

**Comment by gkGaneshR** [comment]
@andyxning Added dot and made the commit. :)

File: cmd/options/options_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	"testing"
+)
+
+type options struct {
+	Nodename         string
+	HostnameOverride string
+}
+
+// TestSetNodeNameOrDie tests for permutations of nodename, hostname and hostnameoverride.
+func TestSetNodeNameOrDie(t *testing.T) {
```
Comment (andyxning) [inline cmd/options/options_test.go]: Thanks for the patient about this PR. @gkGaneshR . I suggest we write the test function like bellow to make it more clear and straightforward. WDYT.

```

// TestSetNodeNameOrDie tests for permutations of nodename, hostname and hostnameoverride.
func TestSetNodeNameOrDie(t *testing.T) {
	hostName, err := os.Hostname()
	if err != nil {
		t.Errorf("Query hostname error: %v", err)
	}

	uts := map[string]struct {
		WantedNodeName string
		Meta           options
	}{
		"Check hostname override only": {
			WantedNodeName: "hostname-override",
			Meta: options{
				Nodename:         "node-name-env",
				HostnameOverride: "hostname-override",
			},
		},
		"Check hostname override and NDDE_NAME env": {
			WantedNodeName: "node-name-env",
			Meta: options{
				Nodename:         "node-name-env",
				HostnameOverride: "",
			},
		},
		"Check hostname override, NODE_NAME env and hostname": {
			WantedNodeName: hostName,
			Meta: options{
				Nodename:         "",
				HostnameOverride: "",
			},
		},
	}

	for desc, ut := range uts {
		err := os.Unsetenv("NODE_NAME")
		if err != nil {
			t.Errorf("Desc: %v. Unset NODE_NAME env error: %v", desc, err)
		}

		if len(ut.Meta.Nodename) != 0 {
			err := os.Setenv("NODE_NAME", ut.Meta.Nodename)
			if err != nil {
				t.Errorf("Desc: %v. Set NODE_NAME env error: %v", desc, err)
			}
		}

		npdOpts := NewNodeProblemDetectorOptions()
		npdOpts.HostnameOverride = ut.Meta.HostnameOverride
		npdOpts.SetNodeNameOrDie()

		if npdOpts.NodeName != ut.WantedNodeName {
			t.Errorf("Desc: %v. Set node name error. Wanted: %v. Got: %v", desc, ut.WantedNodeName, npdOpts.NodeName)
		}
	}
}
```

Comment (gkGaneshR) [reply cmd/options/options_test.go]: I have a doubt. Consider a case, where NODE_NAME env is set already on the system with "abc", when this test is passed successful, NODE_NAME will have "node-name-env". Is it okay? @andyxning

**Comment by gkGaneshR** [comment]
@andyxning Thanks for the help. Raised pr.

**Pull request merged**